public/app/plugins/datasource/loki/backendResultTransformer.ts

```diff
@@ -0,0 +1,100 @@
+import { DataQueryRequest, DataQueryResponse, DataFrame, isDataFrame, FieldType, QueryResultMeta } from '@grafana/data';
+import { LokiQuery, LokiQueryType } from './types';
+import { makeTableFrames } from './makeTableFrames';
+import { formatQuery, getHighlighterExpressionsFromQuery } from './query_utils';
+
+function isMetricFrame(frame: DataFrame): boolean {
+  return frame.fields.every((field) => field.type === FieldType.time || field.type === FieldType.number);
+}
+
+// returns a new frame, with meta merged with it's original meta
+function setFrameMeta(frame: DataFrame, meta: QueryResultMeta): DataFrame {
+  const { meta: oldMeta, ...rest } = frame;
+  // meta maybe be undefined, we need to handle that
+  const newMeta = oldMeta === undefined ? { ...meta } : { ...oldMeta, ...meta };
+  return {
+    ...rest,
+    meta: newMeta,
+  };
+}
+
+function processStreamsFrames(frames: DataFrame[], queryMap: Map<string, LokiQuery>): DataFrame[] {
+  return frames.map((frame) => {
+    const query = frame.refId !== undefined ? queryMap.get(frame.refId) : undefined;
+    const meta: QueryResultMeta = {
+      preferredVisualisationType: 'logs',
+      searchWords: query !== undefined ? getHighlighterExpressionsFromQuery(formatQuery(query.expr)) : undefined,
+    };
+    return setFrameMeta(frame, meta);
+  });
+}
+
+function processMetricInstantFrames(frames: DataFrame[]): DataFrame[] {
+  return frames.length > 0 ? makeTableFrames(frames) : [];
+}
+
+function processMetricRangeFrames(frames: DataFrame[]): DataFrame[] {
+  const meta: QueryResultMeta = { preferredVisualisationType: 'graph' };
+  return frames.map((frame) => setFrameMeta(frame, meta));
+}
+
+// we split the frames into 3 groups, because we will handle
+// each group slightly differently
+function groupFrames(
+  frames: DataFrame[],
+  queryMap: Map<string, LokiQuery>
+): {
+  streamsFrames: DataFrame[];
+  metricInstantFrames: DataFrame[];
+  metricRangeFrames: DataFrame[];
+} {
+  const streamsFrames: DataFrame[] = [];
+  const metricInstantFrames: DataFrame[] = [];
+  const metricRangeFrames: DataFrame[] = [];
+
+  frames.forEach((frame) => {
+    if (!isMetricFrame(frame)) {
+      streamsFrames.push(frame);
+      return;
+    }
+
+    const isInstantFrame = frame.refId != null && queryMap.get(frame.refId)?.queryType === LokiQueryType.Instant;
+    if (isInstantFrame) {
+      metricInstantFrames.push(frame);
+    } else {
+      metricRangeFrames.push(frame);
+    }
+  });
```

Comment:
this is personal preference/nit pick that makes it easier to read for me. Feel free to ignore.
```suggestion
  frames.forEach((frame) => {
    if (!isMetricFrame(frame)) {
      streamsFrames.push(frame);
    } else {
      const isInstantFrame = frame.refId != null && queryMap.get(frame.refId)?.queryType === LokiQueryType.Instant;
      if (isInstantFrame) {
        metricInstantFrames.push(frame);
      } else {
        metricRangeFrames.push(frame);
      }
    }
  });
```

Reply:
yeah, you are right, two returns here make things confusing, applied!